storage: use NodeLiveness in conjunction with gossip for allocator

spencerkimball opened the pull request:
Previously, we were using only how recently a store had gossiped
its store descriptor to determine whether a store was live for the
purposes of allocations and rebalancing. We now additionally
consult the NodeLiveness table, which requires a node be able to
actively heartbeat its liveness system record. Using only gossip
introduced an asymmetry because gossip is considerably more tolerant
of failures.

Simplified the `StorePool` struct to remove the priority queue and
dedicated goroutine.

Fixes #12101

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12178)
<!-- Reviewable:end -->

spencerkimball committed:
storage: use NodeLiveness in conjunction with gossip for allocator

Previously, we were using only how recently a store had gossiped
its store descriptor to determine whether a store was live for the
purposes of allocations and rebalancing. We now additionally
consult the NodeLiveness table, which requires a node be able to
actively heartbeat its liveness system record. Using only gossip
introduced an asymmetry because gossip is considerably more tolerant
of failures.

Simplified the `StorePool` struct to remove the priority queue and
dedicated goroutine.

Fixes #12101

tamird commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator_test.go, line 177 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUJ1_SIAerBQzYuLmu:-KYUJ1_SIAerBQzYuLmv:b3yfsd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/allocator_test.go#L177)):*
> ```Go
> func createTestAllocator(
> 	deterministic bool, useRuleSolver bool,
> ) (*stop.Stopper, *gossip.Gossip, *StorePool, Allocator, *hlc.ManualClock, *mockNodeLiveness) {
> ```

this extra return value is never used - remove it?

---

*[pkg/storage/allocator_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUFDrTnXwBk8LN7n-1:-KYUFDrTnXwBk8LN7n-2:bg0jgk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/allocator_test.go#L198)):*
> ```Go
> 	defer storePool.mu.Unlock()
> 
> 	liveNodeSet := map[roachpb.NodeID]struct{}{}
> ```

nit: can you use `make`? we have conflicting precedent, but I find `make` easier to read when it's an empty literal.

comment applies throughout, but is optional.

---

*[pkg/storage/store_pool.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUExSyIiYj82mWvUwZ:-KYUExSyIiYj82mWvUw_:b-k2ahjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L75)):*
> ```Go
> 
> // StorePoolNodeLivenessTrue is a NodeLivenessFunc which always returns true.
> func StorePoolNodeLivenessTrue(_ roachpb.NodeID, _ time.Time, _ time.Duration) bool {
> ```

this should be in `helpers_test.go`

---

*[pkg/storage/store_pool.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHOu2ISFU4cd6ZZYf:-KYUHOu2ISFU4cd6ZZYg:b-fvk94c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L120)):*
> ```Go
> 	// If there's no descriptor (meaning no gossip ever arrived for this
> 	// store), we can't check the node liveness, so presume this node live.
> 	if sd.desc == nil {
> ```

how can this happen? wouldn't the above `return` catch this? This looks like it might be necessary in tests which operate in unrealistically early time.

---

*[pkg/storage/store_pool.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHsjydqzl8kpAFL0m:-KYUHsjydqzl8kpAFL0n:biit593) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L132)):*
> ```Go
> 	now time.Time, threshold time.Duration, rangeID roachpb.RangeID, nl NodeLivenessFunc,
> ) storeStatus {
> 	// The store must be alive and it must have a descriptor to be considered
> ```

suggest you remove this comment.

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYUHeGcui4kR3Dmzzna:bz01om8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
> ```Go
> 	// The store must be alive and it must have a descriptor to be considered
> 	// alive.
> 	if sd.isDead(now, threshold, nl) || sd.desc == nil {
> ```

this sd.desc check is now redundant with the logic in `isDead`.

---

*[pkg/storage/store_pool.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHkt_EmDTK7dcr8RA:-KYUHkt_EmDTK7dcr8RB:b6acrvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L138)):*
> ```Go
> 	}
> 
> 	// The store must not be throttled from recently declining a reservation.
> ```

this comment is simply not useful.

---

*[pkg/storage/store_pool.go, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHzyk8i2eGIlJS647:-KYUHzyk8i2eGIlJS648:bo7atfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L175)):*
> ```Go
> 	g *gossip.Gossip,
> 	clock *hlc.Clock,
> 	nodeLiveness NodeLivenessFunc,
> ```

s/nodeLiveness/nodeLivenessFn/

---

*[pkg/storage/store_pool.go, line 212 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUI46wUS6NCSFmccGd:-KYUI46wUS6NCSFmccGe:b-z5je0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L212)):*
> ```Go
> 
> 	var buf bytes.Buffer
> 	now := sp.clock.Now().GoTime()
> ```

I think you want `sp.Clock.PhysicalTime()` here.

---

*[pkg/storage/store_pool.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUID48RSWOixyaWpzS:-KYUID48RSWOixyaWpzT:b-1yqyxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L244)):*
> ```Go
> 	detail := sp.getStoreDetailLocked(storeDesc.StoreID)
> 	detail.desc = &storeDesc
> 	detail.lastUpdatedTime = sp.clock.Now().GoTime()
> ```

ditto, `sp.Clock.PhysicalTime()`

---

*[pkg/storage/store_pool.go, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUILR2sR6yiDv7Y7JT:-KYUILR2sR6yiDv7Y7JU:b6pcpeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L287)):*
> ```Go
> 		// time passes without updates from gossip.
> 		detail = newStoreDetail()
> 		detail.lastUpdatedTime = sp.clock.Now().GoTime()
> ```

another: `sp.Clock.PhysicalTime()`

---

*[pkg/storage/store_pool.go, line 314 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIOFvrdxcRfzqLRt2:-KYUIOFvrdxcRfzqLRt3:b-seemdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L314)):*
> ```Go
> 
> 	var deadReplicas []roachpb.ReplicaDescriptor
> 	now := sp.clock.Now().GoTime()
> ```

this one too.

---

*[pkg/storage/store_pool.go, line 327 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIXqLJkJgPxqxevDc:-KYUIXqLJkJgPxqxevDd:b-yorhf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L327)):*
> ```Go
> 			if deadRepl.ReplicaID == repl.ReplicaID {
> 				deadReplicas = append(deadReplicas, repl)
> 				continue outer
> ```

optional: remove `outer` and use `break` here.

---

*[pkg/storage/store_pool.go, line 436 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIa0aTwjEQGUrNAOx:-KYUIa0aTwjEQGUrNAOy:bpzqsgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L436)):*
> ```Go
> 	var storeDescriptors []roachpb.StoreDescriptor
> 
> 	now := sp.clock.Now().GoTime()
> ```

another.

---

*[pkg/storage/store_pool_test.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUFaLAFvdiN1cPtEuo:-KYUFaLB3TKjaGxI470d:b-iaghip) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool_test.go#L62)):*
> ```Go
> }
> 
> func makeMockNodeLiveness(defaultNodeLiveness bool) *mockNodeLiveness {
> ```

s/make/new/

---

*[pkg/storage/store_pool_test.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUGK071EhUgekNCBvB:-KYUGHxEkUpeDK9PVqn6:b-wrw7rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/76d810bf9b8741d08907fdb543b92fb9da1d24a3/pkg/storage/store_pool_test.go#L131)):*
> ```Go
> // TestStorePoolDies ensures that a store is marked as dead after it
> // times out and that it will be revived after a new update is received.
> func TestStorePoolDies(t *testing.T) {
> ```

is there an equivalent test elsewhere now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store_pool.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUQhB1um-CSTR0xOQT:-KYUQhB2FwNHn25PU81x:b-pg125o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L163)):*
> ```Go
> 	mu                          struct {
> 		syncutil.RWMutex
> 		nodeLiveness   NodeLivenessFunc
> ```

Why does this need to be protected by `StorePool.mu`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-:-KYUR-NYDYN70JM4zrQT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
This greatly simplifies the store pool.  Wonderful.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 7 of 8 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-:-KYUewkRafrY3EhcQSKt:b-80dbdo)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator_test.go, line 177 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUJ1_SIAerBQzYuLmu:-KYUfwmOoNfUwjeoTh5N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/allocator_test.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this extra return value is never used - remove it?
</blockquote></details>

Done.

---

*[pkg/storage/allocator_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUFDrTnXwBk8LN7n-1:-KYUdSo_6M_rdNnWAL10:bxh9269) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/allocator_test.go#L198)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: can you use `make`? we have conflicting precedent, but I find `make` easier to read when it's an empty literal.

comment applies throughout, but is optional.
</blockquote></details>

This is the style others have suggested I use in the past, so it's what I use. I only use `make` when I have a capacity. I'm leaving this as is.

---

*[pkg/storage/store_pool.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUExSyIiYj82mWvUwZ:-KYUdrBMuufh71RC0aqS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L75)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this should be in `helpers_test.go`
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHOu2ISFU4cd6ZZYf:-KYUJfHJjkAu5da2tH92:b-wk0whj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how can this happen? wouldn't the above `return` catch this? This looks like it might be necessary in tests which operate in unrealistically early time.
</blockquote></details>

This can happen if the store was requested but hasn't yet gossiped. Since we don't know the descriptor, we can't very well query the node liveness. This absolutely can happen and will happen in the "real" world, not just tests. I believe the comment explains it adequately. 

---

*[pkg/storage/store_pool.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHsjydqzl8kpAFL0m:-KYUdvU3MPc1ChGGUdmg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

suggest you remove this comment.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYUdygjJ8inrJa7mGRX:b-ouoamm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this sd.desc check is now redundant with the logic in `isDead`.
</blockquote></details>

It's not redundant. `isDead` only checks if `sd.desc == nil` if the store's `lastUpdatedTime` isn't past the threshold. This check for `status` is more aggressive about considering a store dead.  I'm just preserving the existing semantics.

---

*[pkg/storage/store_pool.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHkt_EmDTK7dcr8RA:-KYUekQGjjXwNvzwtBbL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment is simply not useful.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUQhB1um-CSTR0xOQT:-KYUemdTUUlzCKZbbNc8:b2nhcq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L163)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why does this need to be protected by `StorePool.mu`
</blockquote></details>

I was expecting to change it in unittests but that didn't materialize. Moved it out of `mu`.

---

*[pkg/storage/store_pool.go, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHzyk8i2eGIlJS647:-KYUf7mmTfoTeRHBPEeo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/nodeLiveness/nodeLivenessFn/
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 212 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUI46wUS6NCSFmccGd:-KYUfIzHESxbmo1UduHD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L212)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think you want `sp.Clock.PhysicalTime()` here.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUID48RSWOixyaWpzS:-KYUfJUYpKTNvb1FVcG5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto, `sp.Clock.PhysicalTime()`
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUILR2sR6yiDv7Y7JT:-KYUfJw140X9aARZSyQs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

another: `sp.Clock.PhysicalTime()`
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 314 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIOFvrdxcRfzqLRt2:-KYUfKHWbN2bVBoN-vT7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L314)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this one too.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 327 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIXqLJkJgPxqxevDc:-KYUfQ4j_BgL7595nfwv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L327)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

optional: remove `outer` and use `break` here.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 436 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUIa0aTwjEQGUrNAOx:-KYUfKu4AOFL_FDZTpvC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

another.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool_test.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUFaLAFvdiN1cPtEuo:-KYUfW7MMqO81mZxDOzV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool_test.go#L62)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/make/new/
</blockquote></details>

Done.

---

*[pkg/storage/store_pool_test.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUGK071EhUgekNCBvB:-KYUfWymdBeWiJnw4hXM:bp2iu37) ([raw file](https://github.com/cockroachdb/cockroach/blob/76d810bf9b8741d08907fdb543b92fb9da1d24a3/pkg/storage/store_pool_test.go#L131)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is there an equivalent test elsewhere now?
</blockquote></details>

Yes. This became redundant when I removed the store pool's goroutine and priority queue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store_pool.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHOu2ISFU4cd6ZZYf:-KYUm1hEUA2-MMCzsJWd:b-vhdyz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L120)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

This can happen if the store was requested but hasn't yet gossiped. Since we don't know the descriptor, we can't very well query the node liveness. This absolutely can happen and will happen in the "real" world, not just tests. I believe the comment explains it adequately. 
</blockquote></details>

Discussed in person - the confusing thing here is that `lastUpdatedTime` gets set  to now even when nothing is known about the store. A comment on that field roughly mirroring the comment on the site of that value being set would be helpful.

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYUmZLZ43TQpirRUg-o:bavae9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

It's not redundant. `isDead` only checks if `sd.desc == nil` if the store's `lastUpdatedTime` isn't past the threshold. This check for `status` is more aggressive about considering a store dead.  I'm just preserving the existing semantics.
</blockquote></details>

I think this logic needs to be folded into `isDead`. Currently there is inconsistent handling of "deadness" in the various callers of `isDead`. I understand you're preserving existing semantics, but the existing semantics seem confused.

Perhaps it would be most appropriate to file a follow-up issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-:-KYUnNWyiSUOYx5wHfmV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store_pool.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHOu2ISFU4cd6ZZYf:-KYVL-edIW_wMGhscfUA:b-sg8mty) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Discussed in person - the confusing thing here is that `lastUpdatedTime` gets set  to now even when nothing is known about the store. A comment on that field roughly mirroring the comment on the site of that value being set would be helpful.
</blockquote></details>

I added some comments.

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYVL4ZFF5xuxcID5MzJ:b9d6vc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think this logic needs to be folded into `isDead`. Currently there is inconsistent handling of "deadness" in the various callers of `isDead`. I understand you're preserving existing semantics, but the existing semantics seem confused.

Perhaps it would be most appropriate to file a follow-up issue.
</blockquote></details>

I'm going to add a TODO. I agree something is fishy here, but this is exactly how the previous code worked.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 1771 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYVV7wRlJE83frezyiK:-KYVV7wRlJE83frezyiL:blq4dd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/841985c2c446b03f6b024bdad95c5c235e96c8a8/pkg/storage/replica_command.go#L1771)):*
> ```Go
> 	pd.Replicated.State.Lease = &lease
> 	pd.Local.leaseMetricsResult = proto.Bool(true)
> 	if l := r.mu.state.Lease; l == nil || l.Replica.StoreID != lease.Replica.StoreID {
> ```

what's this about?

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYVVMH0_4UZXq43jecv:b4a3pps) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I'm going to add a TODO. I agree something is fishy here, but this is exactly how the previous code worked.
</blockquote></details>

Thanks. I'd feel slightly better if you filed an issue as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 7 of 11 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 1771 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYVV7wRlJE83frezyiK:-KYVYQu4vFCFEqZ-Fa-4:b7x9m37) ([raw file](https://github.com/cockroachdb/cockroach/blob/841985c2c446b03f6b024bdad95c5c235e96c8a8/pkg/storage/replica_command.go#L1771)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's this about?
</blockquote></details>

It's a bit of a drive by. Something that should have been done a long time ago. Just limiting unnecessary gossiping which I witnessed due to us naively re-gossiping the node liveness info every time we refreshed the lease. I was running into this while debugging unittests for this change.

---

*[pkg/storage/store_pool.go, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178#-KYUHeGcui4kR3Dmzzn_:-KYV_UjoVVMonPM8-vlt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9e9ceb88af3b35bdfae4c5f5640ea87c5d5606/pkg/storage/store_pool.go#L134)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Thanks. I'd feel slightly better if you filed an issue as well.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 8 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12178)*
<!-- Sent from Reviewable.io -->
